rio_tiler/io/base.py

```diff
@@ -21,6 +21,14 @@ class BaseReader(metaclass=abc.ABCMeta):
     minzoom: int = attr.ib(init=False)
     maxzoom: int = attr.ib(init=False)
 
+    def __enter__(self):
+        """Support using with Context Managers."""
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers."""
+        pass
+
```

Comment:
@geospatial-jeff I know we agreed to removes those (done in https://github.com/cogeotiff/rio-tiler/pull/271/) But after some though I think it's not a great Idea. 

The main reason to remove those was to ease the use for Async but I do feel using BaseReader is not the way forward for this and we should create a `AsyncBaseReader`). 

The major problem by not having `ctx` has a requirement is that a user could create a Reader without them but then it won't be able to use it in most of our other project that use `with reader() as ...`

Reply:
There is a separate `AsyncBaseReader` [here](https://github.com/cogeotiff/rio-tiler/pull/265/files#diff-2b49975a55ad278a0b6c2f8abaedc733R87).  Thinking through this again I agree that these should be here.  Its standard to have post init logic in `__init__` and keep context managers simple.  But because there is no way to execute asynchronous code within `__init__` we have to run post init logic in `__aenter__` and `__aexit__` for the async base class.

So if we are requiring the async base to be a context manager the sync base should also be to keep the interfaces consistent.  I guess there is one difference; the async base won't be usable outside the context manager, but we could provide a class method which people could call:

```python
reader = await AsyncBaseReader.open()
await reader.close()
```

Reply:
🙏 thanks @geospatial-jeff